3.7/buster/Dockerfile

```diff
@@ -84,26 +90,31 @@ RUN set -ex \
 			test_traceback \
 			test_unicode \
 		' \
-	&& make install \
-	&& rm -rf /usr/src/python \
+	; \
+	make install; \
+	cd /; \
+	rm -rf /usr/src/python; \
 	\
-	&& find /usr/local -depth \
+	find /usr/local -depth \
 		\( \
 			\( -type d -a \( -name test -o -name tests -o -name idle_test \) \) \
 			-o \( -type f -a \( -name '*.pyc' -o -name '*.pyo' -o -name '*.a' \) \) \
 			-o \( -type f -a -name 'wininst-*.exe' \) \
 		\) -exec rm -rf '{}' + \
+	; \
 	\
-	&& ldconfig \
+	ldconfig; \
 	\
-	&& python3 --version
+	python3 --version
 
-# make some useful symlinks that are expected to exist
-RUN cd /usr/local/bin \
-	&& ln -s idle3 idle \
-	&& ln -s pydoc3 pydoc \
-	&& ln -s python3 python \
-	&& ln -s python3-config python-config
+# make some useful symlinks that are expected to exist ("/usr/local/bin/python" and friends)
+RUN set -eux; \
+	for src in idle3 pydoc3 python3 python3-config; do \
+		dst="$(echo "$src" | tr -d 3)"; \
+		[ -s "/usr/local/bin/$src" ]; \
+		[ ! -e "/usr/local/bin/$dst" ]; \
+		ln -svT "/usr/local/bin/$src" "/usr/local/bin/$dst"; \
```

Comment:
Would it be possible for these to remain relative links?

Python treats relative links specially:

```

$ AFTER=python@sha256:6c5fd370ff72da06c7c55a0b292812f230342607d257da564e605d54931a4177
$ BEFORE=python@sha256:520eed245023b4059571e9828d06720bb3f53b4200ce372dad715b136ebf8456
$ CMD='ls -ltr /usr/local/bin/python; ln -s /usr/local/ /tmp/foo; /tmp/foo/bin/python -c "import sys; print(sys.base_prefix)"'

$ docker run -it $BEFORE sh -c "${CMD}"
lrwxrwxrwx 1 root root 7 Jan 26 22:04 /usr/local/bin/python -> python3
/tmp/foo

$ docker run -it $AFTER sh -c "${CMD}"
lrwxrwxrwx 1 root root 22 Jan 29 05:59 /usr/local/bin/python -> /usr/local/bin/python3
/usr/local
```

Reply:
I'm definitely willing to go back to a relative link (I don't think it makes much functional difference either way), but I would love to understand more about what the actual use case/breakage is?

(Your example shows a difference, but it's contrived enough that it's hard to understand how/why it might break in a real program. :sweat_smile:)

Reply:
Yes, it is totally contrived :joy:

The more complete situation is:

1. I provide a bunch of python distributions to colleagues on a shared filesystem, and provide one of them as a "pro" version (the one they should ideally use). 
2. Because of the network architecture, there is a single common location on disk that people can access, but this is just a symlink to machine specific mount-points (I'm not clear on the precise reason for the mount-point changes, could be due to the highly segmented network architecture which provides strong network segregation)
3. When somebody uses the "pro" version to create a virtual environment, I want the "base" python to come from the fixed version, not the floating "pro" version (so that when pro changes, their virtual environments continue to work as before)
4. To achieve this, I provide a tool that (automatically) converts the floating version to a fixed version. I do this by looking at the base prefix, and resolving it if it is a symlink, but I have to be careful not to recursively resolve the symlink as it will make the virtual environment not portable due to the different network mount points. I use the fact that the link is relative to prevent filesystem jumps.

As you say, my example was contrived to simplify the story above. As you can imagine, the implementation is a bit delicate, and so I setup some CI testing of it. The tests stopped passing when the image was released with these changes :wink:

Reply:
:grimacing: 

That's all _extremely_ clever (kudos!), but also not exactly a supported use case for this image. :see_no_evil:

I would suggest you fix the symlinks to be what you expect/need after you extract them from the image, but frankly you're going to have to / get to keep all the pieces there. :sweat_smile:

(We could fix this one specific instance of breakage, but I cannot possibly promise we won't break your setup again in the future. :innocent:)

Reply:
This is totally on me, no question. I understand how this works :stuck_out_tongue_winking_eye:.

Irrespective of how I use/abuse the images, the use of relative symlinks is the more consistent approach with how CPython and venv work out of the box, IMO. I propose it because it will result in mildly better images (and I obviously benefit :smile:).